### PR TITLE
Add ds thread pool

### PR DESCRIPTION
**Pull request by jeffdiclemente**
The introduction of a thread pool is meant to offer better performance in cases where declarative services is running on many core machines and the creation of a thread per std::async call is undesirable. 

boost asio was chosen because it is a robust and stable libraries with a compatible software license which provides a thread pool implementation.

### CODE REVIEW

File: compendium/DeclarativeServices/src/SCRActivator.cpp [hunk 1]
```diff
@@ -82,13 +86,18 @@ void SCRActivator::Stop(cppmicroservices::BundleContext context)
...
     // clear bundle registry
     {
       std::lock_guard<std::mutex> l(bundleRegMutex);
       bundleRegistry.clear();
     }
     // clear component registry
     componentRegistry->Clear();
+
+    // There should be no more DS operations which involve a thread pool at this point.
+    threadpool->join();
```
Comment (jeffdiclemente) [inline compendium/DeclarativeServices/src/SCRActivator.cpp:99]: need to guarantee threadpool->join()is called if something above this throws.

**Comment by jeffdiclemente** [comment]
For some reason the travis CI build status is not showing up. It passed with known, unrelated code coverage failure and can be viewed here: https://travis-ci.org/github/CppMicroServices/CppMicroServices/builds/729397726

File: compendium/DeclarativeServices/src/manager/states/CMDisabledState.cpp [hunk 2]
```diff
@@ -41,44 +41,7 @@ CMDisabledState::CMDisabledState()
...
-                   if (ptr) {
-                     std::rethrow_exception(ptr);
-                   }
-                 })
-        .share();
-    return futObj;
-  }
-  // return the stored future in the current enabled state object
-  return currentState->GetFuture();
+  return cm.PostAsyncDisabledToEnabled(currentState);
```
Comment (kevinleeMW) [inline compendium/DeclarativeServices/src/manager/states/CMDisabledState.cpp:44]: love this!

**Review by kevinleeMW — APPROVED** [review]
looked through all the files except the ones under third_party/boost
looks good! love the ScopeGuard changes.

**Review by pelliott-mathworks — APPROVED** [review]
Line 48 of SCRBundleExtension.cpp Just wondering if the threadpool argument to the constructor should be checked too? Same comment/question line 51 of ComponentManagerImpl.cpp

**Comment by jeffdiclemente** [comment]
> Line 48 of SCRBundleExtension.cpp Just wondering if the threadpool argument to the constructor should be checked too? Same comment/question line 51 of ComponentManagerImpl.cpp

sure, i'll add checks

**Pull request merged**